src/modes/ModeEstimation.jl

```diff
@@ -0,0 +1,318 @@
+            setorder!(vi, vn, get_num_produce(vi))
+        else
+            r = vi[vn]
+        end
+    else
+        r = init(dist, spl)
+        push!(vi, vn, r, dist, spl)
+        settrans!(vi, false, vn)
+    end
+	return r, 0
+end
+
+function DynamicPPL.tilde(ctx::OptimizationContext, spl, dist, vn::VarName, inds, vi)
+    if haskey(vi, vn)
+        # Always overwrite the parameters with new ones for `SampleFromUniform`.
+        if spl isa SampleFromUniform || is_flagged(vi, vn, "del")
+            unset_flag!(vi, vn, "del")
+            r = init(dist, spl)
+            vi[vn] = vectorize(dist, r)
+            settrans!(vi, false, vn)
+            setorder!(vi, vn, get_num_produce(vi))
+        else
+            r = vi[vn]
+        end
+    else
+        r = init(dist, spl)
+        push!(vi, vn, r, dist, spl)
+        settrans!(vi, false, vn)
+    end
+    return r, Distributions.logpdf(dist, r)
+end
+
+# observe
+function DynamicPPL.tilde(ctx::OptimizationContext{<:PriorContext}, sampler, right, left, vi)
+    return 0
+end
+
+function DynamicPPL.tilde(ctx::OptimizationContext, sampler, dist, value, vi)
+    return Distributions.logpdf(dist, value)
+end
+
+# dot assume
+function DynamicPPL.dot_tilde(ctx::OptimizationContext{<:LikelihoodContext}, sampler, right, left, vn::VarName, _, vi)
+    vns, dist = get_vns_and_dist(right, left, vn)
+    r = get_and_set_val!(vi, vns, dist, sampler)
+    return r, 0
+end
+
+function DynamicPPL.dot_tilde(ctx::OptimizationContext, sampler, right, left, vn::VarName, _, vi)
+    vns, dist = get_vns_and_dist(right, left, vn)
+    r = get_and_set_val!(vi, vns, dist, sampler)
+    lp = sum(logpdf(dist, r))
+    return r, lp
+end
+
+# dot observe
+function DynamicPPL.dot_tilde(ctx::OptimizationContext{<:PriorContext}, sampler, right, left, vn, _, vi)
+    return 0
+end
+
+function DynamicPPL.dot_tilde(ctx::OptimizationContext{<:PriorContext}, sampler, right, left, vi)
+    return 0
+end
+
+function DynamicPPL.dot_tilde(ctx::OptimizationContext, sampler, right, left, vn, _, vi)
+    vns, dist = get_vns_and_dist(right, left, vn)
+    r = get_and_set_val!(vi, vns, dist, sampler)
+    lp = sum(logpdf(dist, r))
+    return lp
+end
+
+function DynamicPPL.dot_tilde(ctx::OptimizationContext, sampler, dists, value, vi)
+    return sum(Distributions.logpdf.(dists, value))
+end
+
+"""
+    OptimLogDensity{M<:Model,C<:Context,V<:VarInfo}
+
+A struct that stores the log density function of a `DynamicPPL` model.
+"""
+struct OptimLogDensity{M<:Model,C<:AbstractContext,V<:VarInfo}
+    "A `DynamicPPL.Model` constructed either with the `@model` macro or manually."
+    model::M
+    "A `DynamicPPL.AbstractContext` used to evaluate the model. `LikelihoodContext` or `DefaultContext` are typical for MAP/MLE."
+    context::C
+    "A `DynamicPPL.VarInfo` struct that will be used to update model parameters."
+    vi::V
+    "Whether the to evaluate the model with transformed variables."
+    linked::Bool
```

Comment:
I don't think there should be a choice here? `OptimLogDensity` is only used internally, and hence we know that we always optimize in the unconstrained space.

Reply:
I think it's when computing the Hessian that `linked` needs to be false.

Reply:
Turns out we don't need to unlink it for the Hessian, since the OptimizationContext handles everything so nicely.

Reply:
Cool.